x/subscription/keeper/subscription.go

```diff
@@ -0,0 +1,198 @@
+package keeper
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/lavanet/lava/utils"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	packages "github.com/lavanet/lava/x/packages/keeper"
+	"github.com/lavanet/lava/x/subscription/types"
+)
+
+// SetSubscription sets a subscription (of a consumer) in the store
+func (k Keeper) SetSubscription(ctx sdk.Context, sub types.Subscription) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+	b := k.cdc.MustMarshal(&sub)
+	store.Set(types.SubscriptionKey(sub.Consumer), b)
+}
+
+// GetSubscription returns the subscription of a given consumer
+func (k Keeper) GetSubscription(ctx sdk.Context, consumer string) (val types.Subscription, found bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+
+	b := store.Get(types.SubscriptionKey(consumer))
+	if b == nil {
+		return val, false
+	}
+
+	k.cdc.MustUnmarshal(b, &val)
+	return val, true
+}
+
+// RemoveSubscription removes the subscription (of a consumer) from the store
+func (k Keeper) RemoveSubscription(ctx sdk.Context, consumer string) {
+	sub, _ := k.GetSubscription(ctx, consumer)
+	packages.PutPackage(ctx, sub.PackageIndex, sub.Block)
+
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+	store.Delete(types.SubscriptionKey(consumer))
+}
+
+// GetAllSubscription returns all subscriptions that satisfy the condition
+func (k Keeper) GetCondSubscription(ctx sdk.Context, cond func(sub types.Subscription) bool) (list []types.Subscription) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.SubscriptionKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.Subscription
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		if cond == nil || cond(val) {
+			list = append(list, val)
+		}
+	}
+
+	return
+}
+
+// GetAllSubscription returns all subscription (of all consumers)
+func (k Keeper) GetAllSubscription(ctx sdk.Context) []types.Subscription {
+	return k.GetCondSubscription(ctx, nil)
+}
+
+func endOfMonth(date time.Time) time.Time {
```

Comment:
maybe general time-related functions should be under common? (and not in subscription module)